server/svix-server/src/core/idempotency.rs

```diff
@@ -0,0 +1,615 @@
+        IdempotencyKey(base64::encode(&res))
+    }
+}
+
+impl CacheKey for IdempotencyKey {
+    const PREFIX_CACHE: &'static str = "SVIX_IDEMPOTENCY_CACHE";
+}
+
+impl AsRef<str> for IdempotencyKey {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+async fn resolve_service<S>(
+    mut service: S,
+    req: Request<Body>,
+) -> Result<Response<BoxBody>, Infallible>
+where
+    S: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
+    S::Response: IntoResponse,
+    S::Future: Send + 'static,
+{
+    service.call(req).await.map(IntoResponse::into_response)
+}
+
+/// The idempotency middleware itself -- inserted in place of a regular route
+#[derive(Clone)]
+pub struct IdempotencyService<S: Clone> {
+    pub redis: Option<RedisCache>,
+    pub service: S,
+}
+
+impl<S> Service<Request<Body>> for IdempotencyService<S>
+where
+    S: Service<Request<Body>, Error = Infallible> + Clone + Send + 'static,
+    S::Response: IntoResponse,
+    S::Future: Send + 'static,
+{
+    type Response = Response<BoxBody>;
+    type Error = Infallible;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let mut service = self.service.clone();
+        let redis = self.redis.clone();
+
+        if let Some(redis) = redis {
+            Box::pin(async move {
+                let (parts, body) = req.into_parts();
+
+                // If not a POST request, simply resolve the service as usual
+                if parts.method != http::Method::POST {
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                }
+
+                let auth = if let Some(Ok(auth)) =
+                    parts.headers.get("Authorization").map(|v| v.to_str())
+                {
+                    auth
+                } else {
+                    // No auth token -- pass off to service and do not cache
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                };
+
+                let key = if let Some(Ok(key)) =
+                    parts.headers.get("idempotency-key").map(|v| v.to_str())
+                {
+                    key
+                } else {
+                    // No idempotency-key -- pass off to service and do not cache
+                    return resolve_service(service, Request::from_parts(parts, body)).await;
+                };
+
+                let uri = parts.uri.to_string();
+                let key = IdempotencyKey::new(auth, key, &uri);
+
+                // Check Redis now
+                match redis.get::<SerializedResponse>(&key).await {
+                    // If the value is a starting value, the operation is still in progress
```

Comment:
The content of each arm of this match would benefit from being move to its own function the way it is now is very hard to read/follow.